### PR TITLE
[refactor] Move FrontendContext from global into Callable class.

### PR DESCRIPTION
**Pull request by ailzhang**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #2839
* #2837
* #2836
* #2835
* #2834
* #2833

Fixes https://taichi-graphics.atlassian.net/browse/LAN-24

### CODE REVIEW

File: taichi/program/function.cpp [hunk 1]
```diff
@@ -12,10 +12,10 @@ Function::Function(Program *program, const FunctionKey &func_key)
 
 void Function::set_function_body(const std::function<void()> &func) {
   // Do not corrupt the context calling this function here
-  auto backup_context = std::move(taichi::lang::context);
+  auto backup_context = std::move(context);
```
Comment (k-ye) [inline taichi/program/function.cpp:15]: This should no longer be needed...

File: taichi/program/kernel.cpp [hunk 2]
```diff
@@ -22,24 +22,22 @@ Kernel::Kernel(Program &program,
     : grad(grad), lowered_(false) {
   this->program = &program;
   // Do not corrupt the context calling this kernel here -- maybe unnecessary
-  auto backup_context = std::move(taichi::lang::context);
+  auto backup_context = std::move(context);
```
Comment (k-ye) [inline taichi/program/kernel.cpp:25]: Ditto

**Review by k-ye — APPROVED** [review]
This is truly great, thanks!

File: taichi/program/function.cpp (hunk 1 above)
Comment (ailzhang) [reply taichi/program/function.cpp:15]: Nice catch! Thanks!

**Pull request merged**